app/src/main/java/com/karumi/ui/presenter/SuperHeroesPresenter.kt

```diff
@@ -1,22 +1,42 @@
 package com.karumi.ui.presenter
 
+import co.metalab.asyncawait.async
 import com.karumi.ui.LifecycleSubscriber
+import com.karumi.ui.domain.Result.Left
+import com.karumi.ui.domain.Result.Right
+import com.karumi.ui.domain.model.SuperHero
+import com.karumi.ui.domain.usecase.GetSuperHeroes
 import java.lang.ref.WeakReference
 
-class SuperHeroesPresenter(view: View) : LifecycleSubscriber {
+class SuperHeroesPresenter(view: View, getSuperHeroes: GetSuperHeroes) : LifecycleSubscriber {
     private val viewWeak: WeakReference<View> = WeakReference(view)
+    private val getSuperHeroes: GetSuperHeroes = getSuperHeroes
 
     private fun view() = viewWeak.get()
 
     override fun initialize() {
-        view()?.hideLoading()
     }
 
     override fun update() {
+        refreshSuperHeroes()
+    }
 
+    private fun refreshSuperHeroes() {
+        async {
+            val result = await { getSuperHeroes.get() }
+            view()?.hideLoading()
+            when (result) {
+                is Right -> view()?.showSuperHeroes(result.value)
+                is Left -> TODO("Show error")
```

Comment:
We are not handling errors in the original kata. You can just remove this code, show an error view or use the logger here.

Reply:
I added it because this is a sealed class, If i remove it does not fails, but It's good take in mind both cases. I will add an error view in next iteration.

Reply:
Please, add just a log trace. The original implementation does not handle errors because is an extra exercise we request to the training attendants.